### ISSUE TITLE
MRCA function?

### ISSUE THREAD

**Issue by isaacovercast** [EcoJulia/Phylo.jl#40]
Hey folks,
One more question/request. Is there a "good" way to get the mrca node from a list of tip labels? Right now I'm doing this (essentially get all ancestors of all tips and find the shallowest one), and it feels hackish, but I can't figure out the canonical way to do it:

```
target = "substring_of_samples_of_interest"

function get_mrca(tre, target)
    lfs = [x for x in getleaves(tre) if occursin(target, string(x))]
    anc = [getancestors(tre, x) for x in lfs]
    mrca = intersect(anc...)[1]
end
```
Thanks a bunch,
-isaac

**Comment by mkborregaard** [comment]
It would be great to have an MRCA function. I don't think your solution is too hacky, though I think a slightly more efficient algorithm could be to iterate from tips and back, stopping when a recognized species is seen. Something like this
```julia
function find_mrca(tree, target)
    ancestors = getancestors(tree, target[1])
    ranks = Dict(j => i for (i,j) in enumerate(ancestors))
    checked = Set(ancestors)
    oldest = 1
    for species in target
        while !(species ∈ checked)
            push!(checked, species)
            species = getparent(tree, species)
        end
        oldest = max(oldest, get(ranks, species, 0))
    end
    ancestors[oldest]
end
```
use as
```julia
using Phylo
tree = open(parsenewick, "hummingbirds_tree.txt")
target = rand(collect(nodenamefilter(isleaf, tree)), 5)
find_mrca(tree, target)
```
I just wrote this code, haven't validated it against ape etc, but see if you can build something useful from it

**Comment by mkborregaard** [comment]
@richardreeve wouldn't this be quicker if it operated directly on the nodes and not on the node names? But it seems `getparent` only works for names?

**Comment by richardreeve** [comment]
`getparent()` should use whatever it gets as input, so if you pass it a name it'll return a name. Different trees are in principle faster for different inputs, but the default (`LinkTree`s) do use the node objects. To use that here, you just need to call `getnode(first(target))` or to call `nodefilter()` in the first place instead of `nodenamefilter()`.

**Comment by mkborregaard** [comment]
Ah, nice - the code works and just outputs the results in the same type as the elements of `target`!
Using the nodes seems a lot more efficient for my toy example (3 times) but if it varies among trees we could just leave it as it is and let the user decide.

**Comment by mkborregaard** [comment]
OK I've drafted a PR at #41  comments very appreciated.

**Comment by richardreeve** [comment]
There's actually a trait `PreferNodeObjects{T}` which tells you which type to use. I think `getnode()` actually returns the preferred type, because in a really simple tree structure there may be no object that represents the node. `getnodename()` is only there because sometimes you really do want to know the name.

**Comment by richardreeve** [comment]
PS And you're right:
```julia
_prefernodeobjects(::Type{<:LinkNode}) = true
```

**Comment by mkborregaard** [comment]
So is the right design here to ducktype or transform to linknodes internally?

**Comment by richardreeve** [comment]
I'm not fussed - I've been writing `@traitfn`s that convert and convert back if it's not in the "native" format - you can see it in `src/Interface.jl`, but I don't feel strongly about it.

**Comment by mkborregaard** [comment]
I feel like it's better to leave it to the user to keep track of the types then

**Comment by mkborregaard** [comment]
Closed in #41 
@isaacovercast when you think of other key infrastructure functions missing like this one, please open issues - it's a good excuse for adding some more of that.

**Comment by isaacovercast** [comment]
Will do!
